File: GenMotionEasyCpp/src/AxisController.cpp
#include "GenMotionEasyCpp/AxisController.h"
#include "GenMotionEasyCpp/GtnError.h"
#include <memory>
#include <sstream>

/* ---- 子模块实现（内联在 AxisController.cpp 中） ---- */

namespace {

class AxisHomeImpl : public IAxisHome {
    short _core, _axis;
public:
    AxisHomeImpl(short c, short a) : _core(c), _axis(a) {}
    short SetHomeMode(short mode) override { return GTN_SetHomingMode(_core, _axis, mode); }
    short SetHomingParam(short method, double switchSpeed, double indexSpeed,
        double acc, long offset, unsigned short probeFunction) override {
        return GTN_SetEcatHomingPrm(_core, _axis, method, switchSpeed, indexSpeed, acc, offset, probeFunction);
    }
    short StartHoming() override { return GTN_StartEcatHoming(_core, _axis); }
    short GetHomingStatus(unsigned short& status) override {
        return GTN_GetEcatHomingStatus(_core, _axis, &status);
    }
};

class MoveMotionImpl : public IMoveMotion {
    short _core, _axis;
public:
    MoveMotionImpl(short c, short a) : _core(c), _axis(a) {}
    void MoveAbsolute(long pos, double vel, double acc, double dec, short percent) override {
        TMoveAbsolutePrm prm;
        prm.pos = pos; prm.vel = vel; prm.acc = acc; prm.dec = dec; prm.percent = percent;
        GtnException::ThrowIfError(GTN_MoveAbsolute(_core, _axis, &prm), "GTN_MoveAbsolute");
    }
    TMoveAbsolutePrm GetMoveAbsolute() override {
        TMoveAbsolutePrm prm;
        GtnException::ThrowIfError(GTN_GetMoveAbsolute(_core, _axis, &prm), "GTN_GetMoveAbsolute");
        return prm;
    }
    void MoveAbsoluteEx(long pos, double vel, double acc, double dec,
        short percent, double velStart, double velEnd,
        double accStartPercent, double decEndPercent) override {
        TMoveAbsolutePrmEx prm;
        prm.pos = pos; prm.vel = vel; prm.acc = acc; prm.dec = dec;
        prm.percent = percent; prm.velStart = velStart; prm.velEnd = velEnd;
        prm.accStartPercent = accStartPercent; prm.decEndPercent = decEndPercent;
        GtnException::ThrowIfError(GTN_MoveAbsoluteEx(_core, _axis, &prm), "GTN_MoveAbsoluteEx");
    }
    TMoveAbsolutePrmEx GetMoveAbsoluteEx() override {
        TMoveAbsolutePrmEx prm;
        GtnException::ThrowIfError(GTN_GetMoveAbsoluteEx(_core, _axis, &prm), "GTN_GetMoveAbsoluteEx");
        return prm;
    }
    void MoveVelocity(double vel, double acc, double dec, short direction,
        double jerkBegin, double jerkEnd) override {
        TMoveVelocityPrm prm;
        prm.vel = vel; prm.acc = acc; prm.dec = dec; prm.direction = direction;
        prm.jerkBegin = jerkBegin; prm.jerkEnd = jerkEnd;
        GtnException::ThrowIfError(GTN_MoveVelocity(_core, _axis, &prm), "GTN_MoveVelocity");
    }
    void MoveVelocityPositive(double vel, double acc, double dec, double jb, double je) override {
        MoveVelocity(vel, acc, dec, 0, jb, je);
    }
    void MoveVelocityNegative(double vel, double acc, double dec, double jb, double je) override {
        MoveVelocity(vel, acc, dec, 1, jb, je);
    }
    TMoveVelocityPrm GetMoveVelocity() override {
        TMoveVelocityPrm prm;
        GtnException::ThrowIfError(GTN_GetMoveVelocity(_core, _axis, &prm), "GTN_GetMoveVelocity");
        return prm;
    }
};

class JogMotionImpl : public IJogMotion {
    short _core, _axis;
public:
    JogMotionImpl(short c, short a) : _core(c), _axis(a) {}
    void SetJogMode(double acc, double dec, double smooth) override {
        GtnException::ThrowIfError(GTN_PrfJog(_core, _axis), "GTN_PrfJog");
        TJogPrm prm;
        GTN_GetJogPrm(_core, _axis, &prm);
        prm.acc = acc; prm.dec = dec; prm.smooth = smooth;
        GtnException::ThrowIfError(GTN_SetJogPrm(_core, _axis, &prm), "GTN_SetJogPrm");
    }
    void JogMove(double vel) override {
        GTN_SetVel(_core, _axis, vel);
        GTN_Update(_core, 1 << (_axis - 1));
    }
    void Stop() override { GTN_Stop(_core, 1, 1); }
};

class PointMotionImpl : public IPointMotion {
    short _core, _axis;
public:
    PointMotionImpl(short c, short a) : _core(c), _axis(a) {}
    void PointMove(long pos, double vel, double acc, double dec) override {
        TTrapPrm trap;
        GTN_PrfTrap(_core, _axis);
        GTN_GetTrapPrm(_core, _axis, &trap);
        trap.acc = acc; trap.dec = dec; trap.smoothTime = 50;
        GTN_SetTrapPrm(_core, _axis, &trap);
        GTN_SetVel(_core, _axis, vel);
        GTN_SetPos(_core, _axis, pos);
        GTN_Update(_core, 1 << (_axis - 1));
    }
    void PointAbsMove(long pos, double vel, double acc, double dec) override {
        TMoveAbsolutePrmEx move;
        move.acc = acc; move.dec = dec; move.pos = pos; move.vel = vel;
        GTN_MoveAbsoluteEx(_core, _axis, &move);
    }
};

} // anonymous namespace

/* ---- AxisController PIMPL ---- */

class AxisController::Impl {
public:
    Impl(short core, short axis, short crd)
        : _core(core), _axis(axis), _crd(crd)
        , axisHome(std::make_unique<AxisHomeImpl>(core, axis))
        , move(std::make_unique<MoveMotionImpl>(core, axis))
        , jog(std::make_unique<JogMotionImpl>(core, axis))
        , point(std::make_unique<PointMotionImpl>(core, axis))
    {}

    short _core, _axis, _crd;
    std::unique_ptr<AxisHomeImpl> axisHome;
    std::unique_ptr<MoveMotionImpl> move;
    std::unique_ptr<JogMotionImpl> jog;
    std::unique_ptr<PointMotionImpl> point;
};

AxisController::AxisController(short core, short axis, short crd)
    : _core(core), _axis(axis), _crd(crd)
    , _lock(new int())
    , _impl(new Impl(core, axis, crd))
{}

AxisController::~AxisController() = default;

IAxisHome* AxisController::AxisHome() { return _impl->axisHome.get(); }
IMoveMotion* AxisController::Move() { return _impl->move.get(); }
IJogMotion* AxisController::Jog() { return _impl->jog.get(); }
IPointMotion* AxisController::Point() { return _impl->point.get(); }

IPTMotion* AxisController::PT() { return nullptr; }
IFollowMotion* AxisController::Follow() { return nullptr; }
IPVTMotion* AxisController::PVT() { return nullptr; }
IFollowExMotion* AxisController::FollowEx() { return nullptr; }
IGearMotion* AxisController::Gear() { return nullptr; }
IInterpMotion* AxisController::Interp() { return nullptr; }

short AxisController::Restart() { return GTN_Reset(_core); }
short AxisController::LoadConfig(const char* path) { return GTN_LoadConfig(_core, const_cast<char*>(path)); }
short AxisController::EnableAxis() { return GTN_AxisOn(_core, _axis); }
short AxisController::DisableAxis() { return GTN_AxisOff(_core, _axis); }
short AxisController::ClearAlarm(short count) { return GTN_ClrSts(_core, _axis, count); }
void AxisController::StopAxis() { GTN_Stop(_core, 1, 1); }

StatusInfo AxisController::GetEcatStatus() {
    StatusInfo info;
    long sts;
    unsigned long clk;
    if (GTN_GetSts(_core, _axis, &sts, 1, &clk) != 0) return info;

    info.axisAlarm = (sts & 0x2) != 0;
    info.followAlarm = (sts & 0x10) != 0;
    info.plusLimitAlarm = (sts & 0x20) != 0;
    info.minusLimitAlarm = (sts & 0x40) != 0;
    info.smoothStopAlarm = (sts & 0x80) != 0;
    info.scram = (sts & 0x100) != 0;
    info.enableAxis = (sts & 0x200) != 0;
    info.planning = (sts & 0x400) != 0;

    long prfMode;
    GTN_GetPrfMode(_core, _axis, &prfMode, 1, &clk);
    const char* modeNames[] = {"Trap","Jog","PT","Gear","Follow","Interp","PVT"};
    info.motionType = (prfMode >= 0 && prfMode <= 6) ? modeNames[prfMode] : "\u672a\u77e5";

    long encPos;
    double prfPos, encVel, encAcc, prfVel, prfAcc;
    unsigned long dummy;
    GTN_GetEcatEncPos(_core, _axis, &encPos);
    GTN_GetPrfPos(_core, _axis, &prfPos, 1, &dummy);
    GTN_GetEcatEncVel(_core, _axis, &encPos);
    info.driveVel = (double)encPos;
    GTN_GetAxisEncAcc(_core, _axis, &encAcc, 1, &dummy);
    GTN_GetAxisPrfVel(_core, _axis, &prfVel, 1, &dummy);
    GTN_GetAxisPrfAcc(_core, _axis, &prfAcc, 1, &dummy);

    GTN_GetEcatEncPos(_core, _axis, &encPos);
    info.driveLocation = (double)encPos;
    info.plannedLocation = prfPos;
    info.followErr = prfPos - info.driveLocation;
    info.driveAccVel = encAcc;
    info.plannedVel = prfVel;
    info.plannedAccVel = prfAcc;

    return info;
}

long AxisController::GetRemainingDistance(long targetPos) {
    long encPos;
    GTN_GetEcatEncPos(_core, _axis, &encPos);
    return targetPos - encPos;
}

File: GsnMotionEasy/Motion/Control/AxisController.cs
using GsnMotionEasy.Model;
using GsnMotionEasy.Motion.Control.Details;
using GsnMotionEasy.Motion.Interfaces;
using System;
using System.Diagnostics;
using System.Threading.Tasks;
using static GTN.mc;

namespace GsnMotionEasy.Motion.Control
{
    /// <summary>
    /// 运动控制器轴（Gsn / Glink-II 卡）。
    /// 与 GenMotionEasy.AxisController 结构一致，但：
    ///  - 不含 EtherCAT 总线成员（EcatLoad/EcatState/EcatStart）；
    ///  - 状态查询使用非 Ecat 函数（GTN_GetEncPos / GTN_GetAxisEncVel 等）。
    /// </summary>
    public class AxisController : IAxisController
    {
        private readonly short _core;
        private readonly short _axis;
        private readonly object _lock = new object();
        private bool _homingTag = false;

        public IAxisHome AxisHome { get; }
        public IPTMotion PT { get; }
        public IFollowMotion Follow { get; }
        public IPVTMotion PVT { get; }
        public IMoveMotion Move { get; }
        public IFollowExMotion FollowEx { get; }
        public IGearMotion Gear { get; }
        public IInterpMotion Interp { get; }
        public IPointMotion Point { get; }
        public IJogMotion Jog { get; }


        public AxisController(short core, short axis, short crd = 1)
        {
            _core = core;
            _axis = axis;
            AxisHome = new AxisHome(_core, axis, _lock);
            PT = new AxisMotionPT(_core, axis, _lock);
            Follow = new AxisMotionFollow(_core, axis, _lock);
            PVT = new AxisMotionPVT(_core, axis, _lock);
            Move = new AxisMotionMove(_core, axis, _lock);
            FollowEx = new AxisMotionFollowEx(_core, axis, _lock);
            Gear = new AxisMotionGear(_core, axis, _lock);
            Interp = new AxisMotionInterp(_core, crd, _lock);
            Point = new AxisMotionPoint(_core, axis, _lock, () => ClearAlarm(), () => GetStatus(), () => EnableAxis());
            Jog = new AxisMotionJog(_core, axis, _lock);
        }


        public short Restart()
        {
            lock (_lock)
            {
                return GTN_Reset(_core);
            }
        }

        public short LoadConfig(string path)
        {
            lock (_lock)
            {
                return GTN_LoadConfig(_core, path);
            }
        }


        public short EnableAxis()
        {
            lock (_lock)
            {
                return GTN_AxisOn(_core, _axis);
            }
        }


        public short DisableAxis()
        {
            lock (_lock)
            {
                return GTN_AxisOff(_core, _axis);
            }
        }


        public short ClearAlarm(short count = 1)
        {
            lock (_lock)
            {
                return GTN_ClrSts(_core, _axis, count);
            }
        }


        /// <summary>
        /// 查询轴状态（使用非 Ecat 函数：GTN_GetEncPos / GTN_GetAxisEncVel 等）。
        /// </summary>
        public StatusInfo? GetStatus()
        {
            StatusInfo status = new StatusInfo();
            short rtn;
            int tag;
            uint p;
            rtn = GTN_GetSts(_core, _axis, out tag, 1, out p);
            if (rtn != 0) { return null; }
            //伺服报警
            if ((tag & 0x2) != 0)
            {
                status.AxisAlarm = true;
                if (status.AxisAlarm)
                {
                    Console.WriteLine($"{_axis}轴报警触发");
                }
            }
            else
            {
                status.AxisAlarm = false;
            }
            // 跟随误差越限标志
            if ((tag & 0x10) != 0)
            {
                status.FollowAlarm = true;
            }
            else
            {
                status.FollowAlarm = false;
            }

            //平滑停止触发
            if ((tag & 0x80) != 0)
            {
                //触发
                status.SmoothStopAlarm = true;
            }
            else
            {
                status.SmoothStopAlarm = false;
            }
            //急停触发
            if ((tag & 0x100) != 0)
            {
                status.Scram = true;
            }
            else
            {
                status.Scram = false;
            }
            //伺服使能标志
            if ((tag & 0x200) != 0)
            {
                //使能
                status.EnableAxis = true;
            }
            else
            {
                status.EnableAxis = false;
            }
            //规划器正在运动标志
            if ((tag & 0x400) != 0)
            {
                //正在运动
                status.Planning = true;
            }
            else
            {
                status.Planning = false;
            }
            //正限位
            if ((tag & 0x20) != 0)
            {
                status.PlusLimitAlarm = true;
            }
            else
            {
                status.PlusLimitAlarm = false;
            }
            //负限位
            if ((tag & 0x40) != 0)
            {
                status.MinusLimitAlarm = true;
            }
            else
            {
                status.MinusLimitAlarm = false;
            }
            int pValue;
            uint pClock;
            GTN_GetPrfMode(_core, _axis, out pValue, 1, out pClock);
            switch (pValue)
            {
                case 0:
                    status.MotionType = "Trap";
                    break;
                case 1:
                    status.MotionType = "Jog";
                    break;
                case 2:
                    status.MotionType = "PT";
                    break;
                case 3:
                    status.MotionType = "Gear";
                    break;
                case 4:
                    status.MotionType = "Follow";
                    break;
                case 5:
                    status.MotionType = "Interpolation";
                    break;
                case 6:
                    status.MotionType = "PVT";
                    break;
                default:
                    status.MotionType = "未知";
                    break;
            }
            double encValue;
            uint encClock;
            double prfValue;
            uint prfClock;
            GTN_GetEncPos(_core, _axis, out encValue, 1, out encClock);//编码器位置
            GTN_GetPrfPos(_core, _axis, out prfValue, 1, out prfClock);//规划位置
            status.DriveLocation = encValue;
            status.PlannedLocation = prfValue;
            status.FollowErr = prfValue - encValue;
            double encVelValue;
            uint encVelClock;
            GTN_GetAxisEncVel(_core, _axis, out encVelValue, 1, out encVelClock);
            status.DriveVel = encVelValue;
            double encAccValue;
            uint encAccClock;
            GTN_GetAxisEncAcc(_core, _axis, out encAccValue, 1, out encAccClock);
            status.DriveAccVel = encAccValue;
            double prfVelValue;
            uint prfVelClock;
            GTN_GetAxisPrfVel(_core, _axis, out prfVelValue, 1, out prfVelClock);
            status.PlannedVel = prfVelValue;
            double prfVelAccValue;
            uint prfVelAccClock;
            GTN_GetAxisPrfAcc(_core, _axis, out prfVelAccValue, 1, out prfVelAccClock);
            status.PlannedAccVel = prfVelAccValue;
            return status;
        }


        /// <summary>
        /// 查询剩余距离：目标位置 - 编码器位置
        /// </summary>
        public int GetRemainingDistance(int targetPos)
        {
            double encPos;
            uint clock;
            GTN_GetEncPos(_core, _axis, out encPos, 1, out clock);
            return targetPos - (int)encPos;
        }

        /// <summary>
        /// 等待轴到达目标位置并停止运动。
        /// 根据速度与距离估算运动时间，再加上 extraSeconds 秒的缓冲。
        /// </summary>
        /// <param name="targetPos">目标位置</param>
        /// <param name="vel">运动速度</param>
        /// <param name="acc">加速度</param>
        /// <param name="dec">减速度</param>
        /// <param name="tolerance">规划位置与编码器位置允许的偏差</param>
        /// <param name="extraSeconds">额外等待秒数（在估算时间上增加）</param>
        /// <returns>是否在超时前到达</returns>
        public async Task<bool> WaitAxisStop(int targetPos, double vel, double acc, double dec,
            double tolerance = 10, int extraSeconds = 5)
        {
            double encPos;
            uint encClock;
            GTN_GetEncPos(_core, _axis, out encPos, 1, out encClock);
            double distance = Math.Abs(targetPos - encPos);

            double safeVel = Math.Max(vel, 1);
            double safeAcc = Math.Max(acc, 1);
            double safeDec = Math.Max(dec, 1);
            double estimatedTime = distance / safeVel + safeVel / safeAcc + safeVel / safeDec;
            int waitMs = (int)(estimatedTime * 1000) + extraSeconds * 1000;

            var sw = Stopwatch.StartNew();
            while (sw.ElapsedMilliseconds < waitMs)
            {
                double prfPos;
                uint prfClock;
                GTN_GetPrfPos(_core, _axis, out prfPos, 1, out prfClock);

                double encPosNow;
                uint encClockNow;
                GTN_GetEncPos(_core, _axis, out encPosNow, 1, out encClockNow);

                if (Math.Abs(prfPos - encPosNow) <= tolerance && Math.Abs(prfPos - targetPos) <= tolerance)
                {
                    return true;
                }

                await Task.Delay(50);
            }

            return false;
        }


        public void StopAxis()
        {
            lock (_lock)
            {
                GTN_Stop(_core, 1 << (_axis - 1), 1 << (_axis - 1));
            }
        }

    }
}
